### PR TITLE
docs: docstring for `concat`

### PR DESCRIPTION
**Pull request by Sherwin-14**
I wrote the docstring's for concat function. Please have a look into it

### CODE REVIEW

**Review by MarcoGorelli — COMMENTED** [review]
nice one @Sherwin-14 , thanks!

a few requests:
- could we make `data_1`, `data_2` please? and then `df_pd_1`, `df_pd_2`, `df_pl_1`, `df_pl_2`?
- could you check your statement about `Series`? I don't think we support them

**Comment by Sherwin-14** [comment]
Thanks for the feedback, Marc! The CI tests seem to be failing though and I can't figure it out why, any clues?

**Comment by Sherwin-14** [comment]
> * df_pd_1

Hey, needed some clarification over this. I have changed both the data's to `data_1` and `data_2` but I am confused as to how we can change df it to `df_pd_1` or ` df_pl_1` ?

File: narwhals/functions.py [hunk 1]
```diff
@@ -33,6 +33,64 @@ def concat(
...
+        │ --- ┆ --- │
+        │ i64 ┆ i64 │
+        ╞═════╪═════╡
+        │ 1   ┆ 4   │
+        │ 2   ┆ 5   │
+        │ 3   ┆ 6   │
+        │ 5   ┆ 1   │
+        │ 2   ┆ 4   │
+        └─────┴─────┘
+    """
```
Comment (FBruzzesi) [inline narwhals/functions.py:92]: Would you mind adding an example for horizontal concat as well?

**Review by FBruzzesi — COMMENTED** [review]
Thanks for the PR. I have only one comment in addition to what Marco already mentioned.

Regarding the test failing, that's because we compare the exposed functions with what we expose in [`/stable/v1.py`](https://github.com/narwhals-dev/narwhals/blob/main/narwhals/stable/v1.py). I think this was not failing before because the docstring was empty. Now it should be added.

Here a hint on how to:

```py
def concat(
    items: Iterable[FrameT],
    *,
    how: Literal["horizontal", "vertical"] = "vertical",
) -> FrameT:
    # Same docstring but using 
    # import narwhals.stable.v1 as nw 
    # inplace of import narwhals as nw
    
    return _stableify(nw.concat(items, how=how))
```

**Comment by MarcoGorelli** [comment]
> I have changed both the data's to data_1 and data_2 but I am confused as to how we can change df to df_pd_1 or  df_pl_1 ?

you can do
```
data_1 = ...
data_2 = ...
df_pd_1 = pd.DataFrame(data_1)
df_pd_2 = pd.DataFrame(data_2)
df_pl_1 = pl.DataFrame(data_1)
df_pl_2 = pl.DataFrame(data_2)
```

File: narwhals/functions.py [hunk 2]
```diff
@@ -33,6 +33,112 @@ def concat(
...
+        >>> df_pd_2 = pd.DataFrame(data_2)
+        >>> df_pl_1 = pl.DataFrame(data_1)
+        >>> df_pl_2 = pl.DataFrame(data_2)
+
+        Let's define a dataframe-agnostic function:
+
+        >>> @nw.narwhalify
+        ... def func(df1, df2):
+        ...     native_namespace_df1 = nw.get_native_namespace(df1)
+        ...     native_namespace_df2 = nw.get_native_namespace(df2)
```
Comment (MarcoGorelli) [inline narwhals/functions.py:77]: why do you need this?

File: narwhals/functions.py [hunk 3]
```diff
@@ -33,6 +33,112 @@ def concat(
     *,
     how: Literal["horizontal", "vertical"] = "vertical",
 ) -> FrameT:
+    """
+    Concatenate multiple DataFrames, LazyFrames into a single entity.
+
+    Notes:
+        Only horizontal and vertical concatenations are supported.
+
```
Comment (MarcoGorelli) [inline narwhals/functions.py:41]: i think we can just remove this, the arguments section already shows what's supported

File: narwhals/functions.py [hunk 4]
```diff
@@ -33,6 +33,112 @@ def concat(
...
+        >>> df_pd_2 = pd.DataFrame(data_2)
+        >>> df_pl_1 = pl.DataFrame(data_1)
+        >>> df_pl_2 = pl.DataFrame(data_2)
+
+        Defining a dataframe-agnostic function:
+
+        >>> @nw.narwhalify
+        ... def func(df1, df2):
+        ...     native_namespace_df1 = nw.get_native_namespace(df1)
+        ...     native_namespace_df2 = nw.get_native_namespace(df2)
```
Comment (MarcoGorelli) [inline narwhals/functions.py:119]: same

**Review by MarcoGorelli — COMMENTED** [review]
thanks for updating!

File: narwhals/functions.py [hunk 5]
```diff
@@ -33,6 +33,105 @@ def concat(
...
+        │ 2   ┆ 4   │
+        └─────┴─────┘
+
+        Let's look at case a for horizontal concatenation:
+
+        >>> import pandas as pd
+        >>> import polars as pl
+        >>> import narwhals.stable.v1 as nw
+        >>> data_1 = {"a": [1, 2, 3], "b": [4, 5, 6]}
+        >>> data_2 = {"c": [5, 2], "d": [1, 4]}
```
Comment (MarcoGorelli) [inline narwhals/functions.py:102]:
```suggestion
        >>> data_1 = {"a": [1, 2, 3], "b": [4, 5, 6]}
        >>> data_2 = {"c": [5, 2], "d": [1, 4]}
```

**Review by MarcoGorelli — APPROVED** [review]
nice one, thanks @Sherwin-14 !

**Review by MarcoGorelli — CHANGES_REQUESTED** [review]
one more thing (as ci is failing)

can you update the docstring for `concat` in `narwhals/stable/v1` as well please? it should be the same, but with `import narwhals.stable.v1 as nw` instead of `import narwhals as nw`

**Comment by Sherwin-14** [comment]
> one more thing (as ci is failing)
> 
> can you update the docstring for `concat` in `narwhals/stable/v1` as well please? it should be the same, but with `import narwhals.stable.v1 as nw` instead of `import narwhals as nw`

Hey I checked the  `narwhals/stable/v1` file and I did not find concat function there

**Comment by Sherwin-14** [comment]
> > one more thing (as ci is failing)
> > can you update the docstring for `concat` in `narwhals/stable/v1` as well please? it should be the same, but with `import narwhals.stable.v1 as nw` instead of `import narwhals as nw`
> 
> Hey I checked the `narwhals/stable/v1` file and I did not find concat function there

Hey did you find some solution for this? the ci is still failing and I didn't find the `concat` function in `stable/v1/`.

**Comment by FBruzzesi** [comment]
Hey @Sherwin-14 , sorry I missed the first comment among various notifications.

I now see the issue, we somehow have the `concat` in `stable/v1.py` out of sync, or even worse, not even `_stableify`-ed yet.
@MarcoGorelli I think we can do that in another PR as it seems a bit out of scope for this one?

**Comment by MarcoGorelli** [comment]
thanks for the ping, and sorry for the delay

in `v1.py`, `concat` currently just gets imported and re-exported:
```
from narwhals.functions import concat
```

so yes, maybe a separate PR to add a docstring to `concat` and stableify it would be good 👍

**Comment by Sherwin-14** [comment]
No worries. I can do this but what exactly needs to be done here? If you can outline the steps to me it will be a little bit easier for me.

**Comment by MarcoGorelli** [comment]
thanks!

you'll need to do something like
```diff
diff --git a/narwhals/stable/v1.py b/narwhals/stable/v1.py
index 9c7c2d15..d96f0523 100644
--- a/narwhals/stable/v1.py
+++ b/narwhals/stable/v1.py
@@ -37,7 +37,7 @@ from narwhals.expr import Expr as NwExpr
 from narwhals.expr import Then as NwThen
 from narwhals.expr import When as NwWhen
 from narwhals.expr import when as nw_when
-from narwhals.functions import concat
+from narwhals.functions import concat as nw_concat
 from narwhals.functions import show_versions
 from narwhals.schema import Schema as NwSchema
 from narwhals.series import Series as NwSeries
@@ -1668,6 +1668,9 @@ def from_dict(
         nw.from_dict(data, schema=schema, native_namespace=native_namespace)
     )
 
+def concat(*args, **kwargs):
+    return _stableify(nw_concat(*args, **kwargs))
+
 ```
but instead of `*args` and `**kwargs`, put the actual variable names from the `concat` signature

**Comment by MarcoGorelli** [comment]
looks like Francesco's already on it in  #869 though ⚡

**Comment by FBruzzesi** [comment]
Hey @Sherwin-14 , we just merged a PR that adds `concat` function in `stable/v1.py`. Now you should be able to proceed with this PR after merging main branch

**Comment by Sherwin-14** [comment]
Thanks a lot. I will do the needful :)

**Comment by MarcoGorelli** [comment]
thanks @Sherwin-14 ! i think maybe something went wrong when merging, as docs/api-completeness.md is being shown as modified

**Comment by Sherwin-14** [comment]
> thanks @Sherwin-14 ! i think maybe something went wrong when merging, as docs/api-completeness.md is being shown as modified

Yeah I fixed that you can take a look at it now :)

**Review by MarcoGorelli — APPROVED** [review]
this is really good - thanks @Sherwin-14 , and @FBruzzesi for review!

**Comment by Sherwin-14** [comment]
> this is really good - thanks @Sherwin-14 , and @FBruzzesi for review!
> 
>  [ ](https://camo.githubusercontent.com/2f3443f5f57e94310b336cd66d7ca304fdcc6cf79cdbf49e222decf6295927dc/68747470733a2f2f692e666c756666792e63632f364b5678547368784d73727442384d6a5a374d48766c7642434a6a66746774502e676966) 68747470733a2f2f692e666c756666792e63632f364b5678547368784d73727442384d6a5a374d48766c7642434a6a66746774502e676966 [ ](https://camo.githubusercontent.com/2f3443f5f57e94310b336cd66d7ca304fdcc6cf79cdbf49e222decf6295927dc/68747470733a2f2f692e666c756666792e63632f364b5678547368784d73727442384d6a5a374d48766c7642434a6a66746774502e676966)

**Pull request merged**